### PR TITLE
Add support to define (overwrite) of securityContext or pod security context

### LINKED ISSUE #1437: Feat: Add support to define (overwrite) SecurityContext
**Describe the solution you'd like**
One of requirements from our Cyber security team is to drop capabilities where possible to improve security posture (reduce attack surface) for containers running in k8s.

In values.yaml file to define something like (example):
```yaml
...
securityContext:
  capabilities:
    drop:
      - ALL
  runAsNonRoot: true
  allowPrivilegeEscalation: false
  seccompProfile:
    type: RuntimeDefault
```

**Anything else you would like to add:**
At the moment, this is not possible as it's hardcoded https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml#L113

**Environment:**

- Secrets Store CSI Driver version: (use the image tag): v1.4.1
- Kubernetes version: (use `kubectl version`): EKS version 1.26.12

### PR DESCRIPTION
**Pull request by cbugneac-nex**
**What this PR does / why we need it**:
Add support to define (overwrite) SecurityContext for containers & pod security context.

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #1437

**Special notes for your reviewer**:

Link helm chart:
```bash 
$ helm lint ./
==> Linting ./

1 chart(s) linted, 0 chart(s) failed
```

**TODOs**:

- [ ] squashed commits
- [x] includes documentation
- [ ] adds unit tests

### CODE REVIEW

**Comment by cbugneac-nex** [comment]
Hi @aramase @nilekhc please could you review this PR ?

**Comment by cbugneac-nex** [comment]
Hi @aramase @nilekhc @ritazh @tam7t please could you review this PR when you have time ?

File: manifest_staging/charts/secrets-store-csi-driver/values.yaml [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
+  podSecurityContext: {}
+
   image:
     repository: registry.k8s.io/csi-secrets-store/driver
     tag: v1.4.1
     #digest: sha256:
     pullPolicy: IfNotPresent
+  # The secrets-store container security context
+  securityContext:
+    privileged: true
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/values.yaml:14]: `privileged: true` is required for the driver to function correctly. Moving this to `values.yaml` could cause the driver functionality to fail if not set correctly by the user.

`privileged: true` should instead remain in the manifest and this field should append to the existing `securityContext`.

**Review by aramase — CHANGES_REQUESTED** [review]

File: manifest_staging/charts/secrets-store-csi-driver/values.yaml (hunk 1 above)
Comment (cbugneac-nex) [reply manifest_staging/charts/secrets-store-csi-driver/values.yaml]: Thanks for suggestion. Done.

**Comment by cbugneac-nex** [comment]
Hi @aramase please could you review it once again ?

**Comment by cbugneac-nex** [comment]
Hi @aramase @nilekhc @ritazh @tam7t please could you review this PR when you have time ?

File: manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml [hunk 2]
```diff
@@ -36,6 +36,12 @@ spec:
       {{- end }}
       affinity:
 {{ toYaml .Values.linux.affinity | indent 8 }}
+      securityContext:
+        {{- /* This is required for the driver to function correctly. */}}
+        privileged: true
+        {{- with .Values.podSecurityContext }}
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml:42]:
```suggestion
        {{- with .Values.linux.podSecurityContext }}
```

File: manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml [hunk 3]
```diff
@@ -36,6 +36,12 @@ spec:
...
+        {{- toYaml . | nindent 8 }}
+        {{- end }}
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml:44]:
```suggestion
        {{- toYaml .Values.linux.podSecurityContext | nindent 8 }}
```

File: manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml [hunk 4]
```diff
@@ -48,6 +54,10 @@ spec:
             - --csi-address=/csi/csi.sock
             - --kubelet-registration-path={{ .Values.linux.kubeletRootDir }}/plugins/csi-secrets-store/csi.sock
           imagePullPolicy: {{ .Values.linux.registrarImage.pullPolicy }}
+          {{- with .Values.linux.registrar.securityContext }}
+          securityContext:
+            {{- toYaml . | nindent 12 }}
+          {{- end }}
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml:60]:
```suggestion
          {{- if .Values.linux.registrar.securityContext }}
          securityContext:
            {{- toYaml .Values.linux.registrar.securityContext | nindent 12}}
          {{- end }}
```

File: manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml [hunk 5]
```diff
@@ -36,6 +36,12 @@ spec:
       {{- end }}
       affinity:
 {{ toYaml .Values.linux.affinity | indent 8 }}
+      securityContext:
+        {{- /* This is required for the driver to function correctly. */}}
+        privileged: true
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml:41]: Why is `privileged: true` being moved from container to the pod security context?

File: manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml [hunk 6]
```diff
@@ -150,7 +162,11 @@ spec:
           {{- else }}
           image: "{{ .Values.linux.livenessProbeImage.repository }}:{{ .Values.linux.livenessProbeImage.tag }}"
           {{- end }}
-          imagePullPolicy: {{ .Values.linux.livenessProbeImage.pullPolicy }}
+          imagePullPolicy: {{ .Values.linux.livenessProbe.pullPolicy }}
+          {{- with .Values.linux.livenessProbe.securityContext }}
+          securityContext:
+            {{- toYaml . | nindent 12 }}
+          {{- end }}
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml:169]: same comment as above

**Review by aramase — CHANGES_REQUESTED** [review]

**Comment by aramase** [comment]
Closing this due to inactivity. Please reopen the PR after the comments are addressed.

/close

**Pull request closed without merging**